### DISCUSSION TITLE
Cuda OOM when most memory still free

### DISCUSSION THREAD

**Discussion by BrunoLeiberTransfoEnergy** [vladmandic/sdnext#1448] [Q&A] [answered]
I get frequent OOM errors like so:

> OutOfMemoryError: CUDA out of memory. Tried to allocate 2.31 GiB (GPU 0; 24.00 GiB total capacity; 5.48 GiB already
> allocated; 17.13 GiB free; 5.57 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting
> max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF

This happens mostly when using highres fix, and I noticed it happens more when my PC has been generating already some pictures before. Closing/reopening vladmandic doesn't help. I checked also with nvidia-smi that no lingering python executable was holding onto the memory but I can't see anything.
Is there a way to fix it ? Is it maybe due to some sort of memory leak, and in this case is there a way to flush the memory without restarting the PC all the time ?

Start up log is:

> 10:18:35-603679 INFO     Starting SD.Next
> 10:18:35-606678 INFO     Python 3.10.11 on Windows
> 10:18:35-633918 INFO     Version: d858db44 Fri Jun 16 20:46:38 2023 -0400
> 10:18:35-970552 INFO     nVidia CUDA toolkit detected
> 10:18:37-117553 INFO     Torch 2.0.1+cu118
> 10:18:37-129554 INFO     Torch backend: nVidia CUDA 11.8 cuDNN 8700
> 10:18:37-130552 INFO     Torch detected GPU: NVIDIA GeForce RTX 3090 VRAM 24576 Arch (8, 6) Cores 82
> 10:18:37-227555 INFO     Enabled extensions-builtin: ['a1111-sd-webui-lycoris', 'clip-interrogator-ext', 'LDSR', 'Lora',
>                          'multidiffusion-upscaler-for-automatic1111', 'ScuNET', 'sd-dynamic-thresholding',
>                          'sd-extension-system-info', 'sd-webui-agent-scheduler', 'sd-webui-controlnet',
>                          'stable-diffusion-webui-images-browser', 'stable-diffusion-webui-rembg', 'SwinIR']
> 10:18:37-232551 INFO     Enabled extensions: ['adetailer', 'ultimate-upscale-for-automatic1111']
> 10:18:37-233553 INFO     No changes detected: Quick launch active
> 10:18:37-237552 INFO     Extension preload: 0.0s C:\Users\bruno\Documents\sd\vladmandic\extensions-builtin
> 10:18:37-239554 INFO     Extension preload: 0.0s C:\Users\bruno\Documents\sd\vladmandic\extensions
> 10:18:37-249551 INFO     Server arguments: ['--medvram', '--upgrade']
> No module 'xformers'. Proceeding without it.
> 10:18:42-239638 INFO     Libraries loaded
> 10:18:42-241638 INFO     Using data path: C:\Users\bruno\Documents\sd\vladmandic
> 10:18:42-245638 INFO     Available VAEs: C:\Users\bruno\Documents\sd\vladmandic\models\VAE 2
> 10:18:42-256639 INFO     Available models: C:\Users\bruno\Documents\sd\vladmandic\models\Stable-diffusion 24
> 10:18:44-071635 INFO     ControlNet v1.1.224
> ControlNet preprocessor location: C:\Users\bruno\Documents\sd\vladmandic\extensions-builtin\sd-webui-controlnet\annotator\downloads
> 10:18:44-175636 INFO     ControlNet v1.1.224
> Image Browser: ImageReward is not installed, cannot be used.
> [-] ADetailer initialized. version: 23.6.2, num models: 11
> 10:18:46-633899 INFO     Loading UI theme: name=sudeepshouche/minimalist style=Dark
> Running on local URL:  http://127.0.0.1:7860
> 
> To create a public link, set `share=True` in `launch()`.
> 10:18:48-486901 INFO     Local URL: http://127.0.0.1:7860/
> 10:18:48-487900 INFO     Initializing middleware
> 10:18:48-585899 INFO     [AgentScheduler] Task queue is empty
> 10:18:48-586900 INFO     [AgentScheduler] Registering APIs
> Loading weights: C:\Users\bruno\Documents\sd\vladmandic\models\Stable-diffusion\simplyBeautiful_v10.safetensors ━━ 0… -…
>                                                                                                                    GB
> 10:18:49-358902 INFO     Setting Torch parameters: dtype=torch.float16 vae=torch.float16 unet=torch.float16
> LatentDiffusion: Running in eps-prediction mode
> DiffusionWrapper has 859.52 M params.
> Loading weights: C:\Users\bruno\Documents\sd\vladmandic\models\VAE\vae-ft-mse-840000-ema-pruned.ckpt ━━━╺━ 210.6… 0:00:…
>                                                                                                            MB
> 10:18:50-571901 INFO     Applying scaled dot product cross attention optimization
> 10:18:50-584898 INFO     Embeddings: loaded=7 skipped=0
> 10:18:50-589899 INFO     Model loaded in 1.6s (load=0.4s create=0.3s apply=0.3s vae=0.6s)
> 10:18:50-823900 INFO     Model load finished: {'ram': {'used': 5.2, 'total': 31.95}, 'gpu': {'used': 1.27, 'total':
>                          24.0}, 'retries': 0, 'oom': 0} cached=0
> 10:18:50-984900 INFO     Startup time: 13.7s (torch=2.9s gradio=0.8s libraries=1.2s scripts=3.5s onchange=0.8s ui=1.5s
>                          launch=0.4s app-started=0.2s checkpoint=2.2s)

**Comment by vladmandic** [comment, marked as answer]
> Tried to allocate 2.31 GiB 
> 24.00 GiB total capacity; 5.48 GiB already allocated; 17.13 GiB free; 5.57 GiB reserved

yeah, this doesn't make sense - why is torch reporting oom when torch itself says it has sufficient free memory:

only thing that comes to mind if your resulting width/height are some strange numbers that cause math errors. one thing i'd make sure is that target width/height are both divisible by 8 so they end up as valid integers when torch is doing allocations.

this is not a strict requirement, but it seems to have large impact on torch.
for example, i've seen cases where generating 650x650 fails, but 768x768 works just fine.

other than that, this would need to be reported upstream to torch itself.

**Reply by BrunoLeiberTransfoEnergy** [reply]
I checked and it is a multiple of 8...
I googled some more and it seems like it's happening often to people, but people usually "fix" this by lowering the resolution or the batch size. Which is a bit sad since my batch size is already 1, and the resolution is high on purpose (that's the whole point of highres fix). It seems I can't go over 10GiB usage on my RTX3090 no matter what. I hope it somehow gets improved over next version of pytorch, because it's a shame to have so much VRAM but being unable to use it.

**Reply by ETA-22** [reply]
Out of curiosity, have you tried tiled diffusion instead of hires fix? I often find it’s a pretty similar result (but won’t OOM), especially if using controlnet as well.
